### ISSUE TITLE
Repeated runs of stream with empty input hangs

### ISSUE THREAD

**Issue by laurynasl-wix** [zio/zio#2164]
The following program never runs to completion:

```
import zio.ZIO
import zio.stream.ZStream

object Repro extends zio.App {

  def stream = ZStream
    .fromIterable[Int](Seq.empty)
    .partitionEither(i => ZIO.succeed(if (i % 2 == 0) Left(i) else Right(i)))
    .map { case (evens, odds) => evens.mergeEither(odds) }
    .use(_.runCollect)
  
  override def run(args: List[String]): ZIO[Repro.Environment, Nothing, Int] = {
    ZIO.sequence(Range(0, 100).toList.map(i => for {
      res <- stream
      _ <- ZIO.effectTotal(println(i, res))
    } yield ())).map(_ => 0)
  }
}
```

From it's output you can see that it always completes 1 iteration and then hangs after an arbitrary number of iterations. 

The following also hangs, but less reliably (thanks @eamelink): 
```
object Repro extends zio.App {

  def stream =
    ZStream[Int]()
      .partitionEither(_ => ZIO.succeed(Left(true)))
      .map { _._1 }
      .use(_.runCollect)

  override def run(args: List[String]) = {
    ZIO.sequence(Range(0, 1000).toList.map(i => for {
      res <- stream
      _ <- ZIO.effectTotal(println(i, res))
    } yield ())).map(_ => 0)
  }
}
```

**Comment by horned-sphere** [comment]
I have encountered this problem too (and diagnosed it). The issue is actually caused by a race in ZStream#distributedDynamicWith. If a consumer gets registered with the distributor after the stream has terminated it never sees the end of the incoming stream. This is only noticable in derived combinators like partitionEither when you feed them an empty stream.

I have a "fixed" version that I'm using that stores the Take that termintaed the stream in the Ref that holds the queues and hands it out to any consumers that join after the end of the stream. I was a bit wary of suggesting it as a PR as I wasn't sure people would appreciate the semantics of distributedDynamicWith changing.

**Comment by iravid** [comment]
@eamelink thanks for the report, and @horned-sphere thank you for diagnosing!

I’ve not had a chance to look at this yet. Maybe @mschuwalow could take a look?

**Comment by mschuwalow** [comment]
@horned-sphere thanks for diagnosing. I would consider this as a bug and actually have a pr open that fixes this is issue among other things. I would strongly appreciate if you could take a look and ensure this resolves your issue https://github.com/zio/zio/pull/2157

**Comment by horned-sphere** [comment]
@mschuwalow Sorry I'm a bit late on this! The semaphore you added looks like it fixes the issue. I think the test case could be improved though. The running multiple times is a red herring (sometimes it will fail in the first time, it's very context dependent as it's a race). I've found from my tests cases that hit this it happens about 30% of the time but faily randomly. I've come up with this test case that will time out reliably with the old version:

```
testM("Additional subscribers after stream termination should see the end of the stream") {
    val stream : ZStream[Any, Nothing, Either[Unit, Unit]] = ZStream.empty
    val sub = for {
      onEnd <- Promise.make[Nothing, Unit].toManaged(_.interrupt)
      withCompletion = stream.ensuring(onEnd.succeed(()))
      add <- withCompletion.distributedDynamicWith[Nothing, Either[Unit, Unit]](
        1, _ => UIO.succeed(_ => true))
    } yield (onEnd, add)
    sub.use {
      case (onEnd, add) =>
        val consume : ZIO[Any, Nothing, Unit] = add.flatMap {
          case (_, queue) =>
            ZStream.fromQueue(queue).unTake.runDrain
        }
        (consume.fork *> onEnd.await *> consume).map {
          _ => assertCompletes
        }
    }
  }
```

I'm not sure it's worth chaninging anything though as your PR has already been merged.

**Comment by iravid** [comment]
@horned-sphere Please send a PR with that test and I'll happily add it :-)

**Comment by horned-sphere** [comment]
On further thought; it's actually faulty. I'll fix it this evening and make a PR.

**Comment by horned-sphere** [comment]
@iravid I've made a version of it that actually works [here](https://github.com/zio/zio/pull/2273).

**Comment by iravid** [comment]
Closed in #2157
